### ISSUE TITLE
API_EventTimeout

### ISSUE THREAD

**Issue by dschick** [sardana-org/sardana#1417]
Hi all,

I am running Sardana on one machine where I also have a spock client running. Here everything is fine.

When I do now connect a second spock client from another machine to the same MacroServer I get contant info messages in this second spock client

```
Dummy-10       INFO     2020-09-17 16:36:11,616 ampere.sxr.lab:10000.Door_scattering_1.Debug: Activating polling. Reason: API_EventTimeout
Dummy-10       INFO     2020-09-17 16:36:11,630 ampere.sxr.lab:10000.Door_scattering_1.Error: Activating polling. Reason: API_EventTimeout
Dummy-10       INFO     2020-09-17 16:36:11,645 ampere.sxr.lab:10000.Door_scattering_1.Info: Activating polling. Reason: API_EventTimeout
Dummy-10       INFO     2020-09-17 16:36:11,660 ampere.sxr.lab:10000.Door_scattering_1.Input: Activating polling. Reason: API_EventTimeout
Dummy-10       INFO     2020-09-17 16:36:11,673 ampere.sxr.lab:10000.Door_scattering_1.MacroStatus: Activating polling. Reason: API_EventTimeout
Dummy-10       INFO     2020-09-17 16:36:11,684 ampere.sxr.lab:10000.Door_scattering_1.Output: Activating polling. Reason: API_EventTimeout
```

What am I am doing wrong here?
Is this maybe related to too slow network connections?

**Comment by reszelaz** [comment]
Hi Daniel,

I just check the same scenario and I don't get API_EventTimeout.
I have some ideas to debug your issue:

Could you please check which `TANGO_HOST` do you use on both machines? There were problems with the heartbeat events in some previous versions of Tango if there was a mismatch between the `TANGO_HOST` on the server side and on the client side. So to discard this, use exactly the same `TANGO_HOST` on both. In your case I assume this should be: `ampere.sxr.lab:10000` (remember to restart the server after such a change).

Whwnever you have it safely configured try in an ipython console the following:

```
In [2]: import tango

In [3]: d = tango.DeviceProxy("tango://ampere.sxr.lab:10000/Door_scattering_1")

In [4]: d.subscribe_event("Output", tango.EventType.CHANGE_EVENT, tango.utils.EventCallback())
```
Let's see if this gives you the API_EventTImeout (please wait here approx. 20 s).

**Comment by dschick** [comment]
Hi @reszelaz ,

thanks for the quick reply.
So the `TANGO_HOST` is exactly the same on both machines.
Using your proposed example script, I do get the timeout right away.

I am currently running on Tango 9.2.5 , which is why I am warned about:
https://github.com/tango-controls/pytango/issues/307 this issue during Sardana start-up.
I guess you are referring to this one.

So I would need to upgrade to `libtango9.3.2` to get rid of it, right?

Best

Daniel

**Comment by reszelaz** [comment]
Hi Daniel,
The warning at the startup is about another thing (ensure omni thread context manager which should avoid problems when running macros in parallel etc) 
This is not to discourage you from the upgrade. Actually i would recommend running Sardana with Tango 9.3.4 rc5 or higher.
But I think that your problems comes from another thing. Maybe there is some network traffic which can not pass between these machines?
Make a similar test with the motor's position attribute. Try to write/read the position using the Device Proxy object. Then subscribe to the position change event and move the motor from another client, do you receive the change events? I assume you would get API_EventTimeout in the case of the motor as well. 
Paste here the output of the tests, please.

**Comment by dschick** [comment]
I just tried what you proposed. Here is the output:

```
import tango
>>> d = tango.DeviceProxy("tango://ampere.sxr.lab:10000/motor/mbitangomotorctrl/0")
>>> d.position
49.0
>>> d.position = 50
>>> d.position = 49
>>> d.subscribe_event("Position", tango.EventType.CHANGE_EVENT, tango.utils.EventCallback())
2020-09-21 08:29:39.032418 MOTOR/MBITANGOMOTORCTRL/0 POSITION CHANGE [ATTR_VALID] 49.0
1
>>> 2020-09-21 08:29:50.929781 MOTOR/MBITANGOMOTORCTRL/0 POSITION CHANGE [API_EventTimeout] Event channel is not responding anymore, maybe the server or event system is down
2020-09-21 08:29:51.012545 MOTOR/MBITANGOMOTORCTRL/0 POSITION CHANGE [ATTR_VALID] 49.0
2020-09-21 08:29:56.824589 MOTOR/MBITANGOMOTORCTRL/0 POSITION CHANGE [ATTR_CHANGING] 49.000875
2020-09-21 08:30:01.028149 MOTOR/MBITANGOMOTORCTRL/0 POSITION CHANGE [API_EventTimeout] Event channel is not responding anymore, maybe the server or event system is down
2020-09-21 08:30:00.893026 MOTOR/MBITANGOMOTORCTRL/0 POSITION CHANGE [ATTR_CHANGING] 49.862375
2020-09-21 08:30:02.442436 MOTOR/MBITANGOMOTORCTRL/0 POSITION CHANGE [ATTR_VALID] 50.0
2020-09-21 08:30:07.153146 MOTOR/MBITANGOMOTORCTRL/0 POSITION CHANGE [ATTR_CHANGING] 49.99915625
2020-09-21 08:30:11.047415 MOTOR/MBITANGOMOTORCTRL/0 POSITION CHANGE [API_EventTimeout] Event channel is not responding anymore, maybe the server or event system is down
2020-09-21 08:30:10.563056 MOTOR/MBITANGOMOTORCTRL/0 POSITION CHANGE [ATTR_CHANGING] 49.2825625
2020-09-21 08:30:12.746116 MOTOR/MBITANGOMOTORCTRL/0 POSITION CHANGE [ATTR_VALID] 49.0
```

The client is actually a raspi on Wifi or LAN (does not matter) and we have an internal network controlled by an OpenWRT Router, which acts as an DHCP and DNS server. We also use the default firewall configuration of our LinuxMint 19.3 system  that hosts the TangoDB and Sardana. I also tried allowing all incoming and outgoing traffic on the Mint machine without success.

To me it seems, that the event seems to work sometimes but has some timeouts. So could the network simply be too slow?

Best

Daniel

**Comment by reszelaz** [comment]
Hi Daniel,

I doubt it has to do with the slow network. The heartbeat is a simple event which is emitted by the server once every 10 s. It goes via a separate socket (on a different port), this would explain why the change events get arrive to the client. Also from your logs I see that indeed you get the API_EventTimeout every 10s - so neither of the hearbeat events arrive.

I think inially you were on the right track thinking about upgrading Tango. From https://github.com/tango-controls/cppTango/issues/315 we can see that different cases were fixes in Tango >= 9.3 regardingt the API_EventTimeout.
The one which sounds possible with your case is the one about multiple network interfaces - you mentioned that you use WiFi and LAN. Do you have both enabled at the same time? If yes, maybe you could try to disable one of them?

If after ugrade of Tango you still get this problem, then I propose to try to reproduce it with TangoTest DS (I bet it will be reproducible as well) and try asking on https://github.com/tango-controls/TangoTickets.


Do you have both interfaces (LAN and WiFi) active at the same time?

**Comment by reszelaz** [comment]
Sorry I closed it accidentally.

**Comment by dschick** [comment]
okay, disabling the wifi and just using LAN, also disabling the second LAN on the TangoDB Server, does not make a difference.
I guess I will try using the recent tango 9.3.6 release. However, it seems, that not everything is available as debian package (python3-pytango, libtango9-java, ...) and we try to avoid compiling everything ourself.
So maybe I have to wait a bit, in order to try solving the issue.

**Comment by reszelaz** [comment]
At ALBA we have packaged Tango 9.3.4 (I think rc7). We also should have Pytango 9.3.x (compatible with the above Tango). Soon we will update the Tango Java package as well, but it is not strictly necessary to upgrade the Java part in your case. All for stretch and buster and the amd64 platform. In case you are interested we could share it somehow with you.

Also, just to confirm the theory you could try it with conda (I think you should find the 9.3.x packages).

**Comment by dschick** [comment]
thanks for the offer, so tango 9.3.4 is already in the debian experimental channel.
Pytango 9.3.x is would be a nice to have as .deb package or wheel, whereas we are using it heavily on RasPis so anyhow we would need an arm version which we might compile ourselves once.
Good to know, that the java part does not need to be upgraded.

However, I was also wondering if anyone could test the opposite direction, just running Sardana with Tango 9.2.5 in order to reproduce this issue. Maybe you have a virtual machine available to test it.

I can confirm, that we have the same issue with Sardana 2.8 and 3.0.3 both running with Tango & PyTango 9.2.5 libs.

Is there a possibility to supress the timeout messages in the spock console?

**Comment by reszelaz** [comment]
Hi Daniel,

> However, I was also wondering if anyone could test the opposite direction, just running Sardana with Tango 9.2.5 in order to reproduce this issue. Maybe you have a virtual machine available to test it.

Yes, as I mentioned on the follow-up meeting we used it for years at ALBA. But just in case I have repeated the tests with the Tango and PyTango comming with Debian Stretch (9.2.5a+dfsg1-2 and 9.2.0-2 respectively) and I was not able to reproduce it in my environment.

> Is there a possibility to supress the timeout messages in the spock console

I think it is a bit complicated. Taurus relies on this error and enables polling as soon as this error event arrives. I would say that we are not ready to ignore the heartbeat errors.

**Comment by dschick** [comment]
Hi @reszelaz , I just disabled the second network interface on my TangoDB host and did a restart and voila there are no heartbeat timeouts anymore.
Maybe you could also try to emulate a second NIC on your virtual machine to see, if this is reproducable.

However, do you think, this could be solved with the latest Tango version or shall I open an issue with tango-controls?

**Comment by reszelaz** [comment]
Good to hear that you found a way to make it work! Hopeful you could live with that for a while until we find a proper solution. 

> Maybe you could also try to emulate a second NIC on your virtual machine to see, if this is reproducable.

I think I should have an iPC with such a hardware without the need to emulate anything. I will test it whenever I find some time.

**Comment by dschick** [comment]
we actually need to enable the second NIC on the TangoDB Host but currently we do not rely on having to Sardana Clients.
I will put our resulst here when upgrading to Tango 9.3.4

Lookign forward to your test with your iPC

**Comment by dschick** [comment]
I think, I solved the issue without upgrading ;)
Since I had also some other issues with my multiple network interfaces, see here:
https://www.tango-controls.org/community/forum/c/general/other/multiple-network-interfaces/
I played around with the omniORB config.
By limiting the endpoint for omniORB globally to my local network all TangoDS are now only published via this IP-range.

With that change, it seems, that the API_EventTimeout is solved.

Daniel
